File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/Rumor.scala
package ch.epfl.pop.model.network.method

import ch.epfl.pop.json.HighLevelProtocol.RumorFormat
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.Message
import ch.epfl.pop.model.objects.{Channel, PublicKey}
import spray.json._

final case class Rumor(senderPk: PublicKey, rumorId: Int, messages: Map[Channel, Array[Message]]) extends Params {

  override def hasChannel: Boolean = true

  override def hasMessage: Boolean = true
  
  def toJsonString: String = {
    this.toJson.toString
  }

}

object Rumor extends Parsable {

  def apply(senderPk: PublicKey, rumorId: Int, messages: Map[Channel, Array[Message]]): Rumor = {
    new Rumor(senderPk, rumorId, messages)
  }

  override def buildFromJson(payload: String): Rumor = payload.parseJson.asJsObject.convertTo[Rumor]

}

File: be1-go/message/query/method/rumor.go
package method

import (
	"popstellar/message/query"
	"popstellar/message/query/method/message"
)

type ParamsRumor struct {
	SenderID string                     `json:"sender_id"`
	RumorID  int                        `json:"rumor_id"`
	Messages map[string]message.Message `json:"messages"`
}

// Rumor defines a JSON RPC rumor message
type Rumor struct {
	query.Base
	ID     int         `json:"id"`
	Params ParamsRumor `json:"params"`
}
